modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java

```diff
@@ -4826,68 +4778,60 @@ public void awaitPendingTasksFinished() throws IgniteCheckedException {
     }
 
     /**
-     * Reorders list of checkpoint pages and splits them into needed number of sublists according to
+     * Reorders list of checkpoint pages and splits them into appropriate number of sublists according to
      * {@link DataStorageConfiguration#getCheckpointThreads()} and
      * {@link DataStorageConfiguration#getCheckpointWriteOrder()}.
      *
-     * @param cpPagesTuple Checkpoint pages tuple.
-     * @param threads Checkpoint runner threads.
+     * @param cpPages Checkpoint pages with overall count and user pages info.
      */
-    private GridMultiCollectionWrapper<FullPageId> splitAndSortCpPagesIfNeeded(
-        IgniteBiTuple<Collection<GridMultiCollectionWrapper<FullPageId>>, Integer> cpPagesTuple,
-        int threads
+    private GridConcurrentMultiPairQueue<PageMemoryEx, FullPageId> splitAndSortCpPagesIfNeeded(
+        T3<Collection<T2<PageMemoryEx, GridMultiCollectionWrapper<FullPageId>>>, Integer, Boolean> cpPages
     ) throws IgniteCheckedException {
-        FullPageId[] pagesArr = new FullPageId[cpPagesTuple.get2()];
+        Set<T2<PageMemoryEx, FullPageId[]>> cpPagesPerRegion = new HashSet<>();
 
         int realPagesArrSize = 0;
 
-        for (GridMultiCollectionWrapper<FullPageId> colWrapper : cpPagesTuple.get1()) {
-            for (int i = 0; i < colWrapper.collectionsSize(); i++)
-                for (FullPageId page : colWrapper.innerCollection(i)) {
-                    if (realPagesArrSize == pagesArr.length)
-                        throw new AssertionError("Incorrect estimated dirty pages number: " + pagesArr.length);
-
-                    pagesArr[realPagesArrSize++] = page;
-                }
-        }
+        int totalPagesCnt = cpPages.get2();
 
-        FullPageId fakeMaxFullPageId = new FullPageId(Long.MAX_VALUE, Integer.MAX_VALUE);
+        for (T2<PageMemoryEx, GridMultiCollectionWrapper<FullPageId>> regPages : cpPages.get1()) {
+            FullPageId[] pages = new FullPageId[regPages.get2().size()];
 
-        // Some pages may have been replaced, need to fill end of array with fake ones to prevent NPE during sort.
-        for (int i = realPagesArrSize; i < pagesArr.length; i++)
-            pagesArr[i] = fakeMaxFullPageId;
+            int pagePos = 0;
 
-        if (persistenceCfg.getCheckpointWriteOrder() == CheckpointWriteOrder.SEQUENTIAL) {
-            Comparator<FullPageId> cmp = new Comparator<FullPageId>() {
-                @Override public int compare(FullPageId o1, FullPageId o2) {
-                    int cmp = Long.compare(o1.groupId(), o2.groupId());
-                    if (cmp != 0)
-                        return cmp;
+            for (int i = 0; i < regPages.get2().collectionsSize(); i++) {
+                for (FullPageId page : regPages.get2().innerCollection(i)) {
+                    if (realPagesArrSize++ == totalPagesCnt)
```

Comment:
This has to be "++realPagesArrSize", right? Very dangerous

Reply:
good catch, thanks ! :) fixed.

Reply:
no, i think a little bit more and hope i correct here !!

Reply:
I insist you to think one more time. Anyway, what's the point of this check? It validated array overflow in previous code. Now we have smaller array and we write into it without such validation. These are just my thoughts.

Reply:
I don`t understand your suggestion here, you think this code should be replaced ?